### ISSUE TITLE
Compiler API: CompilerOptions#project does nothing?

### ISSUE THREAD

**Issue by dsherret** [microsoft/TypeScript#31730]
**TypeScript Version:** 3.6.0-dev.20190602

**Search Terms:** project configFilePath CompilerOptions

**Code**

The `project` compiler option is part of the public api, but has zero references in the TypeScript code base:

![image](https://user-images.githubusercontent.com/1609021/58766752-27049380-8550-11e9-8e62-239ad13d5543.png)

It seems that the internal `configFilePath` is what's used:

![image](https://user-images.githubusercontent.com/1609021/58766766-56b39b80-8550-11e9-8f91-3c12b7d60ea9.png)

It's important to specify the config file path in certain cases. For example, for type reference directive resolution this is done in *program.ts*:

```ts
const containingDirectory = options.configFilePath ? getDirectoryPath(options.configFilePath) : host.getCurrentDirectory();
```

So if someone runs the code in a different working directory, then it won't correctly resolve types based on tsconfig.json because there's no way to provide the `configFilePath` through the public API (maybe there is, but I haven't found anything and I would expect it to use `project`).

**Expected behavior:**

The `project` property is used. Or perhaps `project` is deprecated and `configFilePath` is surfaced?

**Actual behavior:**

The `configFilePath` property is used.

Supplying this internal property fixes the issue I'm having, but I am trying to not rely on internal api.

If you'd like, I can open a PR to fix this, but I'd just need some quick direction on how you'd like it to be fixed. Thanks!

**Comment by RyanCavanaugh** [comment]
`.project` is used in `tsc.ts`.

The config file API and the commandline API are kind of conflated in the implementation. When you run `tsc --some --cmdline --args`, we make a `CompilerOptions` representing the commandline. This is the same *type* as what we parse from `compilerOptions` in a `tsconfig.json`, but some things can get specified only from the commandline (project) and some things can only be specified from a tsconfig file.

TL;DR `project` is set when you run `tsc -p some/tsconfig.json` but has no meaning in a tsconfig file

**Comment by dsherret** [comment]
@RyanCavanaugh ok that makes sense. Thanks!

So then the `project`'s resolved file path will be eventually be passed down to `parseJsonConfigFileContentWorker`, which sets the internal `configFilePath` option.

I'm going to close this because if I pass the optional filepath parameter to `parseJsonConfigFileContent` then it will set the internal `configFilePath` for me.  I was not passing that argument.

Thanks again!
